File: hard/076_Minimum_Window_Substring/min.go
package main

import "fmt"

func minWindow(s string, t string) string {
	if s == "" || t == "" {
		return ""
	}
	var res string
	var total, minLen int
	tbl := make(map[byte]int)
	for i, _ := range t {
		tbl[t[i]]++
	}
	start := -1
	occurence := make(map[byte]int)
	for i, _ := range s {
		ch := s[i]
		if _, ok := tbl[ch]; !ok {
			continue
		}
		if start == -1 {
			start = i
		}
		occurence[ch]++
		if occurence[ch] <= tbl[ch] {
			total++
		}
		if total != len(t) {
			continue
		}
		for ; start <= i; start++ {
			curLen := i - start + 1
			if minLen == 0 || curLen < minLen {
				minLen = curLen
				res = s[start : i+1]
			}
			ch := s[start]
			if _, ok := tbl[ch]; !ok {
				continue
			}
			if occurence[ch] <= tbl[ch] {
				break
			}
			occurence[ch]--
		}
	}
	return res
}

func testMinWindow(s string, t string) {
	ans := minWindow(s, t)
	fmt.Printf("s %q, t %q, ans %q\n", s, t, ans)
}

func main() {
	testMinWindow("1", "1")           //1
	testMinWindow("1", "2")           //""
	testMinWindow("201213", "123")    //213
	testMinWindow("20120213", "1223") //20213
}

File: hard/076_Minimum_Window_Substring/min.py
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        tbl = {}
        for ch in t:
            tbl[ch] = tbl.get(ch, 0) + 1
        start, counter = 0, len(t)
        minStart, minLen = -1, len(s)+1
        for i in xrange(len(s)):
            ch = s[i]
            if ch not in tbl:
                continue
            tbl[ch] -= 1
            if tbl[ch] >= 0:
                counter -= 1
            while start <= i and counter == 0:
                winLen = i-start+1
                if winLen < minLen:
                    minStart, minLen = start, winLen
                ch_start = s[start]
                if ch_start in tbl:
                    if tbl[ch_start] >= 0:
                        counter += 1
                    tbl[ch_start] += 1
                start += 1
        if minStart == -1:
            return ""
        else:
            return s[minStart:minStart+minLen]
